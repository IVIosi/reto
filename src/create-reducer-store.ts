import {useReducer} from "react"

type Reducer<S> = (state: S, payload: unknown) => S

type Reducers<S> = {
  [action: string]: Reducer<S>
}

type Action<R extends Reducers<unknown>> = {
  type: keyof R
  payload: unknown
}

export function createReducerStore<S, R extends Reducers<S>>(initialState: S, reducers: R) {
  return function() {
    const [state, dispatch] = useReducer((state, action: Action<R>) => {
      return reducers[action.type](state, action.payload)
    }, initialState)
    return {
      state,
      dispatch,
    }
  }
}
