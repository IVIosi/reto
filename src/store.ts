import {Context, default as React} from 'react'
import {Container} from './container'

export interface Store<P extends unknown[] = any[], V = any> {
  (...args: P): V
  defaultValue?: V
  optional?: boolean
  displayName?: string
  // defaultProps?
  
  Context?: Context<Container<V>>
}

export type StoreP<S extends Store> = S extends Store<infer P, any> ? P : never
export type StoreV<S extends Store> = S extends Store<any, infer V> ? V : never

export function getStoreContext<S extends Store>(S: S) {
  if (!S.Context) {
    S.Context = React.createContext(new Container<StoreV<S>>(S.defaultValue))
  }
  return S.Context
}
