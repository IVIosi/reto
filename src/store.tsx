import * as React from 'react'
import {contextSymbol} from './metadata-symbols'
import produce, {createDraft, Draft, finishDraft} from 'immer'

export type ConstructorType<T> = { new (...args: any[]): T }
export type StoreType<T extends Store = Store> = ConstructorType<T>

export class Store<T extends {} = {}> {
  state: T
  // getters: {}
  storeWillDestroy?(): void
  mutate(f: (draft: Draft<T>) => void) {
    this.state = produce(this.state, f)
    this.propagate()
  }
  propagate() {
    for (let subscriber of this.subscribers) {
      subscriber()
    }
  }
  subscribers: Function[] = []
  effects = [
    {
      handler: function () {
      
      },
      deps: function () {
      
      }
    }, {
    },
  ]
}


export function store<T extends {new(...args:any[]):{}}>(constructor: T) {
  Reflect.defineMetadata(contextSymbol, React.createContext(null), constructor)
}


// function useFunctionStore(Store: FunctionStore) {
//   const state = Store.init()
// }
