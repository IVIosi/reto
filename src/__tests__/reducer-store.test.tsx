import {Consumer, Provider, useStore, withProvider} from '..'
import {BarStore, FooStore} from './stores/foo.store'
import * as React from 'react'
import {FC, useState} from 'react'
import * as testing from '@testing-library/react'
import {createReducerStore} from '../create-reducer-store'

test('reducer store', function () {
  const MyStore = createReducerStore({
    counter: 1
  }, {
    increase: (state, {amount}) => {
      return {
        ...state,
        counter: state.counter + amount
      }
    }
  })

  const App: FC = (props) => {
    const myStore = useStore(MyStore)

    function changeStore() {
      myStore.dispatch({
        type: 'increase',
        payload: {amount: 2}
      })
    }
    return (
      <div>
        <button onClick={changeStore}>Change</button>
        {myStore.state.counter}
      </div>
    )
  }
  const renderer = testing.render(
    <Provider of={MyStore}>
      <App/>
    </Provider>
  )
  expect(renderer.asFragment()).toMatchSnapshot()
  testing.fireEvent.click(testing.getByText(renderer.container, 'Change'))
  expect(renderer.asFragment()).toMatchSnapshot()
})
