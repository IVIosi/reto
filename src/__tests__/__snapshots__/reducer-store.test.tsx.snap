// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer store 1`] = `
<DocumentFragment>
  <div>
    <button>
      Change
    </button>
    1
  </div>
</DocumentFragment>
`;

exports[`reducer store 2`] = `
<DocumentFragment>
  <div>
    <button>
      Change
    </button>
    3
  </div>
</DocumentFragment>
`;
